src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -34456,6 +34440,379 @@ public class Derived : Base
+                );
+        }
+
+        [Fact]
+        public void UnannotatedParam_MaybeNull_UpdatesArgumentState()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+
+public class Program
+{
+    void M0([MaybeNull] string s0) { }
+
+    void M1(string s1)
+    {
+        M0(s1);
+        _ = s1.ToString(); // 1
+    }
+}
+";
+            var comp = CreateNullableCompilation(new[] { source, MaybeNullAttributeDefinition });
+            comp.VerifyDiagnostics(
+                // (11,13): warning CS8602: Dereference of a possibly null reference.
+                //         _ = s1.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s1").WithLocation(11, 13)
+                );
+        }
+
+        [Fact]
+        public void UnannotatedParam_MaybeNullWhen_UpdatesArgumentState()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+
+public class Program
+{
+    bool M0([MaybeNullWhen(true)] string s0) => false;
+
+    void M1(string s1)
+    {
+        M0(s1);
+        _ = s1.ToString(); // 1
+    }
+
+    void M2(string s1)
+    {
+        _ = M0(s1)
+            ? s1.ToString() // 2
+            : s1.ToString();
+    }
+}
+";
+            var comp = CreateNullableCompilation(new[] { source, MaybeNullWhenAttributeDefinition });
+            comp.VerifyDiagnostics(
+                // (11,13): warning CS8602: Dereference of a possibly null reference.
+                //         _ = s1.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s1").WithLocation(11, 13),
+                // (17,15): warning CS8602: Dereference of a possibly null reference.
+                //             ? s1.ToString() // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s1").WithLocation(17, 15)
+                );
+        }
+
+        [Fact]
+        public void AnnotatedParam_DisallowNull_UpdatesArgumentState()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+
+public class Program
+{
+    void M0([DisallowNull] int? i) { }
+
+    void M1(int? i1)
+    {
+        M0(i1); // 1
+        _ = i1.Value.ToString();
+    }
+}
+";
+            var comp = CreateNullableCompilation(new[] { source, DisallowNullAttributeDefinition });
+            comp.VerifyDiagnostics(
+                // (10,12): warning CS8607: A possible null value may not be used for a type marked with [NotNull] or [DisallowNull]
+                //         M0(i1); // 1
+                Diagnostic(ErrorCode.WRN_DisallowNullAttributeForbidsMaybeNullAssignment, "i1").WithLocation(10, 12)
+                );
+        }
+
+        [Fact]
+        public void UnconstrainedTypeParam_DoesNotUpdateArgumentState()
+        {
+            var source = @"
+public class Program<T>
+{
+    void M0(T t) { }
+
+    void M1()
+    {
+        T t = default;
+        M0(t); // 1
```

Comment:
It feels like it would be illustrative here to try calling `M0(t)` a second time. I expect the state to get upgraded from "maybe-default" to "maybe-null" and for there to be no diagnostic on the second call. Does that match your expectation @jcouv?

Reply:
> I expect the state to get upgraded from "maybe-default" to "maybe-null"

I didn't follow.
I could see that either (1) the state is unchanged (which I believe is the case with this PR), or (2) the state would be set to not-null (ie. assuming that the invocation throws).

---
In reply to: [460329898](https://github.com/dotnet/roslyn/pull/46072#discussion_r460329898) [](ancestors = 460329898)

Reply:
Hmm. I think I inferred a principle that doesn't quite apply to unconstrained generics. My thinking is that if you call a method with a value it's not meant to accept, we update the argument state in a way that avoids warnings on subsequent calls. Thus if you call some method with an unconstrained generic parameter (requiring a maybe-null or not-null state), and you pass an argument with a maybe-default state, we change the argument state to maybe-null.

This is based on a conception of the flow states where:
- "maybe-default" means a variable of a generic type *may contain null, even if the type argument disallows it*.
- "maybe-null" means a variable of a generic type *may contain null if the type argument allows it*.
- "not-null" means the variable doesn't contain null.

I think this is not something to change for this PR, but I will modify the test to record this behavior.